ext/git.c

```diff
@@ -0,0 +1,339 @@
+    while (len > 0 && (str[len - 1] == '\n' || str[len - 1] == '\r')) {
+        len--;
+    }
+    str[len] = '\0';
+    return len;
+}
+
+void normalize_string(zend_string *str) {
+    ZSTR_LEN(str) = remove_trailing_newline(ZSTR_VAL(str));
+}
+
+zend_string *read_git_file(const char *path, bool persistent) {
+    FILE *file = fopen(path, "r");
+    if (!file) {
+        return NULL;
+    }
+
+    char buffer[256];
+    size_t len = fread(buffer, 1, sizeof(buffer) - 1, file);
+    fclose(file);
+
+    if (len == 0) {
+        return NULL;
+    }
+
+    buffer[len] = '\0';
+    len = remove_trailing_newline(buffer);
+    return zend_string_init(buffer, len, persistent);
+}
+
+zend_string *get_commit_sha(const char *git_dir) {
+    char head_path[PATH_MAX];
+#ifndef _WIN32
+    snprintf(head_path, sizeof(head_path), "%s/HEAD", git_dir);
+#else
+    snprintf(head_path, sizeof(head_path), "%s\\HEAD", git_dir);
+#endif
+
+    zend_string *head_content = read_git_file(head_path, false);
+    if (!head_content) {
+        return NULL;
+    }
+
+    const char *ref_prefix = "ref: ";
+    if (strncmp(ZSTR_VAL(head_content), ref_prefix, strlen(ref_prefix)) == 0) {
+        char ref_path[PATH_MAX];
+#ifndef _WIN32
+        snprintf(ref_path, sizeof(ref_path), "%s/%s", git_dir, ZSTR_VAL(head_content) + strlen(ref_prefix));
+#else
+        snprintf(ref_path, sizeof(ref_path), "%s\\%s", git_dir, ZSTR_VAL(head_content) + strlen(ref_prefix));
+#endif
+        zend_string_release(head_content);
+        return read_git_file(ref_path, true);
+    }
+
+    return head_content;
+}
+
+zend_string *get_repository_url(const char *git_dir) {
+    char config_path[PATH_MAX];
+    snprintf(config_path, sizeof(config_path), "%s/config", git_dir);
+
+    FILE *file = fopen(config_path, "r");
+    if (!file) {
+        return NULL;
+    }
+
+    char buffer[256];
+    zend_string *result = NULL;
+    const char *url_prefix = "url = ";
+
+    while (fgets(buffer, sizeof(buffer), file)) {
+        if (strncmp(buffer, "[remote \"origin\"]", 17) == 0) {
+            while (fgets(buffer, sizeof(buffer), file)) {
+                if (buffer[0] == '[') break;
+                char *url = strstr(buffer, url_prefix);
+                if (url) {
+                    result = zend_string_init(url + strlen(url_prefix), strlen(url) - strlen(url_prefix), 1);
+                    normalize_string(result);
+                    break;
+                }
+            }
+            break;
+        }
+    }
+
+    fclose(file);
+    return result;
+}
+
+zend_string *find_git_dir(const char *start_dir) {
+    char current_dir[PATH_MAX];
+    snprintf(current_dir, sizeof(current_dir), "%s", start_dir);
+
+    while (true) {
+        char git_dir[PATH_MAX];
+#ifndef _WIN32
+        snprintf(git_dir, sizeof(git_dir), "%s/.git", current_dir);
+#else
+        snprintf(git_dir, sizeof(git_dir), "%s\\.git", current_dir);
+#endif
+
+        if (access(git_dir, F_OK) == 0) {
```

Comment:
You may choose to handle EPERM differently from other errors? Like short-circuit here. If you're not allowed to access, it means you're in a git dir already, just, no git metadata for you.